crates/sui-package-dump/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+        .context("Failed to fetch max page size")?
+        .service_config
+        .max_page_size)
+}
+
+/// Read all the packages between `after_checkpoint` and `before_checkpoint`, in batches of
+/// `page_size` from the `client` connected to a GraphQL service.
+///
+/// If `after_checkpoint` is not provided, packages will be read from genesis. If
+/// `before_checkpoint` is not provided, packages will be read until the latest checkpoint.
+///
+/// Returns the latest checkpoint that was read from in this fetch, and a list of all the packages
+/// that were read.
+async fn fetch_packages(
+    client: &Client,
+    page_size: i32,
+    after_checkpoint: Option<u64>,
+    before_checkpoint: Option<u64>,
+) -> Result<(Option<u64>, Vec<packages::MovePackage>)> {
+    let packages::Query {
+        checkpoint: checkpoint_viewed_at,
+        packages:
+            packages::MovePackageConnection {
+                mut page_info,
+                mut nodes,
+            },
+    } = client
+        .query(packages::build(
+            page_size,
+            None,
+            after_checkpoint.map(UInt53),
+            before_checkpoint.map(UInt53),
+        ))
+        .await
+        .with_context(|| "Failed to fetch page 1 of packages.")?;
+
+    for i in 2.. {
+        if !page_info.has_next_page {
+            break;
+        }
+
+        let packages = client
+            .query(packages::build(
+                page_size,
+                page_info.end_cursor,
+                after_checkpoint.map(UInt53),
+                before_checkpoint.map(UInt53),
+            ))
+            .await
+            .with_context(|| format!("Failed to fetch page {i} of packages."))?
+            .packages;
+
+        nodes.extend(packages.nodes);
+        page_info = packages.page_info;
+
+        info!(
+            "Fetched page {i} ({} package{} so far).",
+            nodes.len(),
+            if nodes.len() == 1 { "" } else { "s" },
+        );
+    }
+
+    use packages::Checkpoint as C;
+    let last_checkpoint = match (checkpoint_viewed_at, before_checkpoint) {
+        (
+            Some(C {
+                sequence_number: UInt53(v),
+            }),
+            Some(b),
+        ) if b > 0 => Some(v.min(b - 1)),
+        (
+            Some(C {
+                sequence_number: UInt53(c),
+            }),
+            _,
+        )
+        | (_, Some(c)) => Some(c),
+        _ => None,
+    };
+
+    Ok((last_checkpoint, nodes))
+}
+
+/// Write out `pkg` to the `output_dir`ectory, using the package's address and name as the directory
+/// name. The following files are written for each directory:
+///
+/// - `object.bcs` -- the BCS serialized form of the `Object` type containing the package.
+///
+/// - `linkage.json` -- a JSON serialization of the package's linkage table, mapping dependency
+///   original IDs to the version of the dependency being depended on and the ID of the object
+///   on-chain that contains that version.
```

Comment:
```suggestion
///   on chain that contains that version.
```
nit but for consistency